### PR TITLE
Implement #13219: Ability to not shuffle masters connections

### PR DESCRIPTION
**Pull request by Vovan-VE**
| Q             | A
| ------------- | ---
| Is bugfix?    | no
| New feature?  | yes
| Breaks BC?    | ~yes?~ no
| Tests pass?   | ~(no tests)~ ~yes, but HHVM hanged up~ yes
| Fixed issues  | #13219

### CODE REVIEW

**Comment by Vovan-VE** [comment]
Cannot check which master was selected. `Connection::getSlave()` returns another `Connection`, so I can check its `$dsn`. But `open()` only gets `$pdo` from selected master.

**Comment by Vovan-VE** [comment]
Should I add `private $_master` to `Connection` in the same manner as `private $_slave`? Shoul I do it in this PR or in separate?

**Comment by samdark** [comment]
This PR.

File: framework/db/Connection.php [hunk 1]
```diff
@@ -120,7 +120,8 @@
  * read-only.
  * @property Schema $schema The schema information for the database opened by this connection. This property
  * is read-only.
- * @property Connection $slave The currently active slave connection. Null is returned if there is slave
+ * @property Connection $master The currently active master connection. Null is returned if there is no master
```
Comment (Vovan-VE) [inline framework/db/Connection.php]: Is it possible to add `@since 2.0.12` to such property?

Comment (cebe) [inline framework/db/Connection.php]: no, the annotation should be on the getter. it will be shown in the apidocs correctly.

**Comment by Vovan-VE** [comment]
Can this PR be merged into 2.0.11? If so, I will update `@since`s and CHANGELOG.

**Review by SilverFire — APPROVED** [review]
Yes, I think we can accept this PR for 2.0.11 release.

CHANGELOG must be updated

File: framework/db/Connection.php [hunk 2]
```diff
@@ -923,13 +961,30 @@ public function useMaster(callable $callback)
...
+     * Opens the connection to a server in the pool.
+     * This method implements the load balancing among the given list of the servers.
+     * Connections will be tried in sequental order.
+     * @param array $pool the list of connection configurations in the server pool
+     * @param array $sharedConfig the configuration common to those given in `$pool`.
+     * @return Connection the opened DB connection, or `null` if no server is available
+     * @throws InvalidConfigException if a configuration does not specify "dsn"
+     * @since 2.0.11
+     */
+    protected function openFromPoolSequentally(array $pool, array $sharedConfig)
```
Comment (samdark) [inline framework/db/Connection.php]: Sequentally → Sequentially

File: framework/db/Connection.php [hunk 3]
```diff
@@ -923,13 +961,30 @@ public function useMaster(callable $callback)
...
     protected function openFromPool(array $pool, array $sharedConfig)
     {
+        shuffle($pool);
+        return $this->openFromPoolSequentally($pool, $sharedConfig);
+    }
+
+    /**
+     * Opens the connection to a server in the pool.
+     * This method implements the load balancing among the given list of the servers.
+     * Connections will be tried in sequental order.
```
Comment (samdark) [inline framework/db/Connection.php]: Sequental → Sequential

**Review by samdark — CHANGES_REQUESTED** [review]

File: framework/db/Connection.php [hunk 4]
```diff
@@ -364,6 +366,12 @@ class Connection extends Component
      * ```
      */
     public $masterConfig = [];
+    /**
+     * @var bool whether to shuffle [[masters]] before getting one.
+     * @since 2.0.11
+     * @see masters
+     */
+    public $randomizeMasters = true;
```
Comment (samdark) [inline framework/db/Connection.php]: How about renaming it to `shuffleMasters`?

**Review by samdark — APPROVED** [review]

File: tests/framework/db/sqlite/ConnectionTest.php [hunk 5]
```diff
@@ -90,6 +96,41 @@ public function testMasterSlave()
         }
     }
 
+    public function testMasterSlaveRandomization()
```
Comment (dynasource) [inline tests/framework/db/sqlite/ConnectionTest.php]: this test looks to complex and will be hard  to maintain in the future. Cant you refactor it to less code?

**Review by dynasource — CHANGES_REQUESTED** [review]

**Review by cebe — APPROVED** [review]

**Comment by cebe** [comment]
Merged, Thank you!

**Pull request merged**